tests/rptest/tests/upgrade_test.py

```diff
@@ -213,46 +243,97 @@ def stop_producer():
+            return True
+
+        # We may start from a version that doesn't even have a logical version, such
+        # as v21.11.x
+        old_logical_version = -1
+
+        while len(self.versions) != 0:
+            self.install_next()
+
+            # Skip producing during upgrade when upgrading to 22.1.x, because this is where
+            # the consumer offsets migration happens.
+            produce_during_upgrade = self.current_version[0:2] != (22, 1)
+            self.logger.info(
+                f"Installed {self.current_version}, doing rolling upgraded (produce during={produce_during_upgrade})"
+            )
+
+            if produce_during_upgrade:
+                # Give ample time to restart, given the running workload.
+                self.installer.install(self.redpanda.nodes,
+                                       self.current_version)
+                self.redpanda.rolling_restart_nodes(self.redpanda.nodes,
+                                                    start_timeout=90,
+                                                    stop_timeout=90)
+
+                # Wait for cluster version to stabilize
+            else:
+                # Upgrading from a pre-22.x Redpanda, so it does not have maintenance
+                # mode, and must go through migration of consumer offsets.
+                stop_producer()
+                self.installer.install(self.redpanda.nodes,
+                                       self.current_version)
+                self.redpanda.rolling_restart_nodes(self.redpanda.nodes,
+                                                    start_timeout=90,
+                                                    stop_timeout=90,
+                                                    use_maintenance_mode=False)
+
+                # When upgrading from versions that don't support maintenance mode
+                # (v21.11 and below), there is a migration of the consumer offsets
+                # topic to be mindful of.
+                rpk = RpkTool(self.redpanda)
+
+                def _consumer_offsets_present():
+                    try:
+                        rpk.describe_topic("__consumer_offsets")
+                    except Exception as e:
+                        if "Topic not found" in str(e):
+                            return False
+                    return True
+
+                wait_until(_consumer_offsets_present,
+                           timeout_sec=90,
+                           backoff_sec=3)
+                self._producer.start(clean=False)
+
+            # After doing a rolling restart with the new version, let the cluster's
+            # logical version and associated feature flag state stabilize.  This avoids
+            # upgrading "too fast" such that the cluster thinks we skipped a version.
+            self.redpanda.wait_until(
+                lambda: logical_version_stable(old_logical_version),
+                timeout_sec=30,
+                backoff_sec=1)
```

Comment:
Interesting -- I wonder if there are k8s hooks or other restart hooks that we can plug into to do these checks in real systems, especially larger systems where it's harder to predict how long a cluster takes to upgrade.